# bwa samtools and Varscan on all .fastq files. For extracting .gz files you'll need to uncomment the first few lines in the align_and_snp_call.R
print("COMMAND: Rscript align_and_snp_call.R")
system("Rscript align_and_snp_call.R")

# Generate the annotation file from ptt and rnt files
print("COMMAND: Rscript annotation_from_ptt_rnt.R")
system("Rscript annotation_from_ptt_rnt.R")

# Generate the gene lengths files (gene lengths with overlapping regions removed)
print ("COMMAND: python annotation_from_ptt_rnt.py")
system ("python annotation_from_ptt_rnt.py")

# Generating snp lists and indel lists
print("COMMAND: python condense_snp_indel_info.py")
system("python condense_snp_indel_info.py") # NEEDS mg1655.annotation FILE GENERATED BY annotation_from_ptt_rnt.R

# Check N and Y reads in all fastq files. Y reads are those that (yes) failed the illumina quality filter. These are the bad reads.
print("COMMAND: python check_NY_reads.py > NY_reads_in_fastq_files.txt")
system("python check_NY_reads.py > NY_reads_in_fastq_files.txt")

# Get a summary of readcounts based on mapping qualities for all files
print("COMMAND: python mapq_readcounts.py > mapq_all_files.txt")
system("python mapq_readcounts.py > mapq_all_files.txt")

# Get the number of reads that are unmapped (i.e mapq 0 and map pos 0)
print("COMMAND: python unmapped_reads.py > unmapped_reads_per_sample.txt")
system("python unmapped_reads.py > unmapped_reads_per_sample.txt")

# Get read counts per nucleotide that are above mapping quality 20
print("COMMAND: python read_count_per_nt.py")
system("python read_count_per_nt.py")

# Get read counts mapping to either the forward strand or reverse strand
print("COMMAND: python per_gene_coding_noncoding_read_counts.py")
system("python per_gene_coding_noncoding_read_counts.py")

print("COMMAND: Rscript check_forward_reverse_corr.R")
system("Rscript check_forward_reverse_corr.R")

# Get read counts per gene # Needs the files generated by the read counts per nucleotide step.
print("COMMAND: python read_count_per_gene.py") 
system("python read_count_per_gene.py") # NEEDS ptt and rnt files

# Combine gene counts into one table using python
print("COMMAND: python tabulate_per_gene_readcounts.py")
system("python tabulate_per_gene_readcounts.py")

